modules/contracts/src.ts/services/ethReader.ts

```diff
@@ -843,6 +962,24 @@ export class EthereumChainReader implements IVectorChainReader {
     return Result.ok(cleaned);
   }
 
+  private async getSafeBlockNumber(chainId: number): Promise<Result<number, ChainError>> {
```

Comment:
We should think about caching this in memory, it's gonna be crazy to constantly query this on every request.

Reply:
We could do add a block listener